src/compute/src/render/top_k.rs

```diff
@@ -300,7 +402,7 @@ where
 
             negated_output
                 .negate()
-                .concat(&input)
+                .concat(&input.map(move |((k, h), r)| ((k, h), R::ok(r))))
```

Comment:
I think this is one of the main things to evaluate about this approach.  I think it keeps the code cleaner, but at this additional cost (plus the trickiness of the errors also going through this negation).  How about this permutation, which is I think a little less clear but not too much so?

```diff
diff --git a/src/compute/src/render/top_k.rs b/src/compute/src/render/top_k.rs
index daa03a2f3..37b45397d 100644
--- a/src/compute/src/render/top_k.rs
+++ b/src/compute/src/render/top_k.rs
@@ -264,10 +266,12 @@ where
             G: Scope,
             G::Timestamp: Lattice,
         {
-            if validating {
+            let input = collection.map(move |((key, hash), row)| ((key, hash % modulus), row));
+            let (oks, errs) = if validating {
                 let stage = build_topk_stage_inner::<G, Result<Row, Row>>(
-                    collection, order_key, modulus, offset, limit, arity,
+                    &input, order_key, offset, limit, arity,
                 );
+
                 let debug_name = debug_name.to_string();
                 let (oks, errs) =
                     stage.map_fallible("Demuxing Errors", move |((k, h), result)| match result {
@@ -284,12 +288,16 @@ where
                 (oks, Some(errs))
             } else {
                 (
-                    build_topk_stage_inner::<G, Row>(
-                        collection, order_key, modulus, offset, limit, arity,
-                    ),
+                    build_topk_stage_inner::<G, Row>(&input, order_key, offset, limit, arity),
                     None,
                 )
-            }
+            };
+            (
+                oks.negate()
+                    .concat(&input)
+                    .consolidate_named::<RowKeySpine<_, _, _>>("Consolidated TopK"),
+                errs,
+            )
         }
 
         // To provide a robust incremental orderby-limit experience, we want to avoid grouping
@@ -301,9 +309,8 @@ where
         // updates, and to reduce the amount of work on the critical path for updates. The cost is
         // a larger number of arrangements when this optimization does nothing beneficial.
         fn build_topk_stage_inner<G, R>(
-            collection: Collection<G, ((Row, u64), Row), Diff>,
+            input: &Collection<G, ((Row, u64), Row), Diff>,
             order_key: Vec<mz_expr::ColumnOrder>,
-            modulus: u64,
             offset: usize,
             limit: Option<usize>,
             arity: usize,
@@ -313,10 +320,9 @@ where
             G::Timestamp: Lattice,
             R: MaybeValidatingRow<Row, Row>,
         {
-            let input = collection.map(move |((key, hash), row)| ((key, hash % modulus), row));
             // We only want to arrange parts of the input that are not part of the actual output
             // such that `input.concat(&negated_output.negate())` yields the correct TopK
-            let negated_output = input
+            input
                 .arrange_named::<RowSpine<(Row, u64), _, _, _>>("Arranged TopK input")
                 .reduce_abelian::<_, RowSpine<_, _, _, _>>("Reduced TopK input", {
                     move |_key, source, target: &mut Vec<(R, Diff)>| {
@@ -373,37 +379,32 @@ where
                         // the `offset` and `limit` constraints.
                         for index in indexes.into_iter() {
                             let (row, mut diff) = source[index];
+                            if !diff.is_positive() {
+                                continue;
+                            }
+                            // If we are still skipping early records ...
+                            if offset > 0 {
+                                let to_skip = std::cmp::min(offset, usize::try_from(diff).unwrap());
+                                offset -= to_skip;
+                                diff -= Diff::try_from(to_skip).unwrap();
+                            }
+                            // We should produce at most `limit` records.
+                            // TODO(benesch): avoid dangerous `as` conversion.
+                            #[allow(clippy::as_conversions)]
+                            if let Some(limit) = &mut limit {
+                                diff = std::cmp::min(diff, Diff::try_from(*limit).unwrap());
+                                *limit -= diff as usize;
+                            }
+                            // Output the indicated number of rows.
                             if diff > 0 {
-                                // If we are still skipping early records ...
-                                if offset > 0 {
-                                    let to_skip =
-                                        std::cmp::min(offset, usize::try_from(diff).unwrap());
-                                    offset -= to_skip;
-                                    diff -= Diff::try_from(to_skip).unwrap();
-                                }
-                                // We should produce at most `limit` records.
-                                // TODO(benesch): avoid dangerous `as` conversion.
-                                #[allow(clippy::as_conversions)]
-                                if let Some(limit) = &mut limit {
-                                    diff = std::cmp::min(diff, Diff::try_from(*limit).unwrap());
-                                    *limit -= diff as usize;
-                                }
-                                // Output the indicated number of rows.
-                                if diff > 0 {
-                                    // Emit retractions for the elements actually part of
-                                    // the set of TopK elements.
-                                    target.push((R::ok(row.clone()), -diff));
-                                }
+                                // Emit retractions for the elements actually part of
+                                // the set of TopK elements.
+                                target.push((R::ok(row.clone()), -diff));
                             }
                         }
                     }
                 })
-                .as_collection(|k, v| (k.clone(), v.clone()));
-
-            negated_output
-                .negate()
-                .concat(&input.map(move |((k, h), r)| ((k, h), R::ok(r))))
-                .consolidate_named::<RowKeySpine<_, _, _>>("Consolidated TopK")
+                .as_collection(|k, v| (k.clone(), v.clone()))
         }
 
         fn render_top1_monotonic<G>(
```

Reply:
Ah, and by the way, the same pattern applies also to the min/max aggregate variant!

Reply:
Oof, thanks, I forgot.